Guide accordion mobile menu

nick-mon1 opened the pull request:
## Summary

Replaces horizontal guide menu to be vertical on mobile. Implements the mobile menu with `usa-accordion` to improve accessibility and mobile UX.

[Link to Preview](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-guide-mobile-menu/guides/hcd/discovery-concepts/)

<details><summary>Screenshots</summary>
<p>

**Before**
<img width="412" alt="Screenshot 2024-04-16 at 9 07 29 AM" src="https://github.com/GSA/digitalgov.gov/assets/104778659/169aa669-79a5-4db7-846f-d72c2632e57d">


**After**
<img width="485" alt="Screenshot 2024-04-30 at 10 47 30 AM" src="https://github.com/GSA/digitalgov.gov/assets/104778659/c6100e8e-cb98-401d-9141-8b47edbec747">

</p>
</details> 


### Solution

Creates a new `guide-menu-links.html` partial that generates the anchor tag links for both desktop and mobile menus.
Creates two guide "sibling" components:

1. `.dg-guide__menu-bar` for desktop
2. `.guide-mobile-menu` for mobile

`guide-mobile-menu.js` will toggle between these menus.

<details><summary>Markup Changes</summary>
<p>

**old markup**

```html
<div class="dg-guide__menu-bar">
  <div class="dg-guide__menu-bar-image-container" hidden>
    <a href="{{- .Site.BaseURL -}}" title="{{- .Site.Title }} Home">
      <picture class="dg-guide__menu-bar-image">
        <source srcset="img/digitalgov-logo.svg" />
        <img src="img/digit-100.svg" />
      </picture>
    </a>
    {{- if $glossary -}}
      <button title="Show glossary" class="dg-glossary__toggle">
        <svg class="dg-glossary__toggle-icon">
          <path
            d="M13.9583 2.99612C13.9583 2.25376 13.5274 1.631 12.9074 1.25803C12.2875 0.885127 11.4706 0.755653 10.6759 0.999207L7.41667 1.99802L4.15746 0.999208C3.36272 0.755654 2.54584 0.885127 1.92596 1.25803C1.30597 1.631 0.875 2.25376 0.875 2.99612V11.6299C0.875 12.5372 1.54014 13.3298 2.50921 13.6268L7.41667 15.1307L12.3241 13.6268C13.2932 13.3298 13.9583 12.5372 13.9583 11.6299V2.99612ZM11.1429 2.17693C11.4882 2.07112 11.8416 2.12807 12.107 2.28774C12.3723 2.44735 12.5417 2.7033 12.5417 2.99612V11.6299C12.5417 11.9855 12.2789 12.3198 11.8571 12.4491L8.125 13.5928V3.10179L11.1429 2.17693ZM2.29167 2.99612C2.29167 2.7033 2.46099 2.44735 2.7263 2.28774C2.99172 2.12808 3.34516 2.07112 3.69046 2.17694L6.70833 3.10179V13.5928L2.97621 12.4491C2.55441 12.3198 2.29167 11.9855 2.29167 11.6299V2.99612Z"
          />
        </svg>
      </button>
    {{- end -}}
  </div>
  <nav class="dg-guide__menu-bar-links" aria-label="Secondary menu bar">
    {{- if $guideData.nav -}}
      {{- range $guideData.nav -}}
        {{- $item := . -}}
        {{- $page := $.Site.GetPage $item.path -}}
        {{- $current := false -}}
        {{- if eq $currentPage $page.RelPermalink -}}
          {{- $current = true -}}
        {{- end -}}
        {{- if $item.subnav -}}
          {{- range $item.subnav -}}
            {{- $subItem := . -}}
            {{- $subPage := $.Site.GetPage $subItem.path -}}
            {{- if eq $currentPage $subPage.RelPermalink -}}
              {{- $current = true -}}
            {{- end -}}
          {{- end -}}
        {{- end -}}
        <a class="dg-guide__menu-bar-link" >{{- $item.title -}}</a>
      {{- end -}}
    {{- end -}}
  </nav>
</div>
```


**new markup**

```html
<div class="dg-guide__menu-bar">
  <div class="dg-guide__menu-bar-image-container" hidden>
    <a href="{{- .Site.BaseURL -}}" title="{{- .Site.Title }} Home">
      <picture class="dg-guide__menu-bar-image">
        <source
          srcset="{{- "img/digitalgov-logo.svg" | relURL -}}"
          media="(min-width: 480px)"
        />
        <img
          src="{{- "img/digit-100.svg" | relURL -}}"
          alt="Digital Gov Logo"
        />
      </picture>
    </a>
    {{- if $glossary -}}
      <button title="Show glossary" class="dg-glossary__toggle">
        <svg class="dg-glossary__toggle-icon">
          <path
            d="M13.9583 2.99612C13.9583 2.25376 13.5274 1.631 12.9074 1.25803C12.2875 0.885127 11.4706 0.755653 10.6759 0.999207L7.41667 1.99802L4.15746 0.999208C3.36272 0.755654 2.54584 0.885127 1.92596 1.25803C1.30597 1.631 0.875 2.25376 0.875 2.99612V11.6299C0.875 12.5372 1.54014 13.3298 2.50921 13.6268L7.41667 15.1307L12.3241 13.6268C13.2932 13.3298 13.9583 12.5372 13.9583 11.6299V2.99612ZM11.1429 2.17693C11.4882 2.07112 11.8416 2.12807 12.107 2.28774C12.3723 2.44735 12.5417 2.7033 12.5417 2.99612V11.6299C12.5417 11.9855 12.2789 12.3198 11.8571 12.4491L8.125 13.5928V3.10179L11.1429 2.17693ZM2.29167 2.99612C2.29167 2.7033 2.46099 2.44735 2.7263 2.28774C2.99172 2.12808 3.34516 2.07112 3.69046 2.17694L6.70833 3.10179V13.5928L2.97621 12.4491C2.55441 12.3198 2.29167 11.9855 2.29167 11.6299V2.99612Z"
          />
        </svg>
      </button>
    {{- end -}}
  </div>
  <nav class="dg-guide__menu-bar-links" aria-label="Secondary menu bar">
    {{ partial "core/guides/guide-menu-links.html" (dict "guideData" $guideData "currentPage" $currentPage "pageScope" .) }}
  </nav>
</div>
```


</p>
</details> 


### How to Test
- [ ] resize browser below 480px to trigger mobile guide menu (vertical navigation)
- [ ] resize browser over 480px to trigger desktop guide menu (horizontal navigation)
- [ ] test on mobile phone browser via [browserstack](https://live.browserstack.com/dashboard)
  - [ ] chrome
  - [ ] safari
  - [ ] firefox


### What to Test
- [ ] **sections +** button will display current guide sections as list items, current guide section is highlighted in blue
- [ ] glossary icon will hide/show glossary on HCD guides
- [ ] digital.gov icon will link back to homepage



| guides                                                                         |
| ------------------------------------------------------------------------------ |
| [accessibility-for-teams](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-guide-mobile-menu/guides/accessibility-for-teams/) |
| [dap](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-guide-mobile-menu/guides/dap/)                                         |
| [hcd](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-guide-mobile-menu/guides/hcd/)                                         |
| [mobile-principles](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-guide-mobile-menu/guides/mobile-principles/)             |
| [public-participation](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-guide-mobile-menu/guides/public-participation/)       |
| [rpa](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-guide-mobile-menu/guides/rpa/)                                         |
| [site-scanning](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-guide-mobile-menu/guides/site-scanning/)                     |
| [web-analytics-playbook](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-guide-mobile-menu/guides/web-analytics-playbook/)   |

---

### Dev Checklist

- [x] PR has correct labels
- [x] A11y testing (voice over testing, meets WCAG, run axe tools)
- [x] Tested against [USWDS accordion testing guidelines](https://designsystem.digital.gov/components/accordion/accessibility-tests/)
- [x] Code is formatted properly

nick-mon1 commented on themes/digital.gov/src/scss/new/guides/_guide-menu-bar.scss:
```diff
@@ -78,3 +78,101 @@
+  }
+
+  &__button {
+    background-color: color("blue-warm-20v");
```
The `usa-accordion__button` uses:

```css
background-image: url(../uswds/img/usa-icons/remove.svg), linear-gradient(transparent, transparent);
```

Have not found a solution to override this hard-coded svg graphic for the +/- icon.

nick-mon1 commented on themes/digital.gov/src/scss/new/guides/_guide-menu-bar.scss:
```diff
@@ -78,3 +78,101 @@
   }
 }
+
+.dg-accordion-mobile-menu {
```
The `dg-accordion-mobile-menu` overrides the `usa-accordion` component styles to work with the guide menu icons and nav list.

nick-mon1 commented on themes/digital.gov/layouts/partials/core/guides/guide-menu-bar.html:
```diff
@@ -65,3 +65,86 @@
+    class="usa-accordion__content dg-accordion-mobile-menu__content usa-prose"
+  >
+    <nav class="dg-accordion-mobile-menu__nav" aria-label="Secondary menu bar">
+      <ul class="dg-accordion-mobile-menu__nav-list">
+        {{- if $guideData.nav -}}
+          {{- range $guideData.nav -}}
+            {{- $item := . -}}
+            {{- $page := $.Site.GetPage $item.path -}}
+            {{- $current := false -}}
+            {{- if eq $currentPage $page.RelPermalink -}}
+              {{- $current = true -}}
+            {{- end -}}
+            {{- if $item.subnav -}}
+              {{- range $item.subnav -}}
+                {{- $subItem := . -}}
+                {{- $subPage := $.Site.GetPage $subItem.path -}}
+                {{- if eq $currentPage $subPage.RelPermalink -}}
+                  {{- $current = true -}}
+                {{- end -}}
+              {{- end -}}
+            {{- end -}}
+            <li class="dg-accordion-mobile-menu__nav-list-item">
+              <a
+                class="dg-accordion-mobile-menu__link {{ cond $current "dg-current" "" }}"
+                href="{{- $page.Permalink -}}"
+                title="{{- $item.title -}}"
+                >{{- $item.title -}}</a
+              >
+            </li>
+          {{- end -}}
+        {{- end -}}
```
This logic for the nav list for the mobile menu is tightly coupled to the markup. I think we could move it out to another partial but need some advice on creating a more generic mobile menu component perhaps as future enhancement.

nick-mon1 commented on themes/digital.gov/src/js/guide-mobile-menu.js:
This code should still work for a generic component when we create one.

nick-mon1 commented on themes/digital.gov/src/scss/new/guides/_guide-menu-bar.scss:
```diff
@@ -78,3 +78,101 @@
+    padding: 0;
+  }
+
+  &__heading {
+    align-items: center;
+    background-color: color("gray-cool-5");
+    display: flex;
+    flex-direction: row;
+    height: 3.25rem;
```
I'm seeing a bug on iOS mobile. Testing with Browsterstack works for Chrome, Firefox and Edge (not pictured).

<img width="924" alt="Screenshot 2024-04-18 at 10 43 32 AM" src="https://github.com/GSA/digitalgov.gov/assets/104778659/f78e5d4d-75c1-4f50-b2c3-fb8f00d59535">

nick-mon1 replied:
Fixed this by setting `width: auto` to `dg-accordion-mobile-menu__image-glossary` and `dg-accordion-mobile-menu__image-glossary svg`.

mejiaj commented on themes/digital.gov/src/js/guide-mobile-menu.js:
```diff
@@ -0,0 +1,65 @@
+ * and style the sections button with token blue-warm-20v
+ * @param {object} event
+ */
+function handleMouseEvent(event) {
+  if (event.type === "mouseenter") {
+    dgAccordionHeader.style.backgroundColor = "#dfe1e2";
+    dgAccordionButton.style.backgroundColor = "#adcdff";
+  } else if (event.type === "mouseleave") {
+    dgAccordionHeader.style.backgroundColor = "#edeff0";
+    dgAccordionButton.style.backgroundColor = "#adcdff";
+  }
+}
```
We should use classes & CSS to control this so we can use USWDS tokens.

mejiaj commented on themes/digital.gov/src/js/guide-mobile-menu.js:
```diff
@@ -0,0 +1,65 @@
+/**
+ * Listen for mouse enter/leave on accordion button to toggle background color to match for the icon
+ */
+function toggleBackgroundColor() {
+  dgAccordionButton.addEventListener("mouseenter", handleMouseEvent);
+  dgAccordionButton.addEventListener("mouseleave", handleMouseEvent);
+}
```
Consider adding these at component initialization. Otherwise these listeners get added multiple times.

mejiaj commented on themes/digital.gov/src/scss/new/guides/_guide-menu-bar.scss:
```diff
@@ -78,3 +79,107 @@
+    }
+  }
+
+  &__nav {
+    &-list {
+      display: flex;
+      flex-direction: column;
+      margin: 0;
+      padding: units(2);
+    }
+
+    &-list-item {
+      align-items: center;
+      display: flex;
+      height: units(6);
+      list-style-type: none;
+    }
+  }
```
Let's avoid nesting too deeply if we can help it. Makes debugging harder.

mejiaj reviewed (changes requested):
Looking good so far. 

Curious about the decision to move the accordion trigger to the left side. I didn't think it looked bad on the right and its placement seemed consistent.

**Mockup with proposed changes**
![image](https://github.com/GSA/digitalgov.gov/assets/3385219/e199935f-6adc-4bde-bbc3-48dea44fcfbb)

Added suggestions based on your comments and also made some styling tweaks. Mainly to match the menu trigger and its dropdown (similar to main menu). As well as applying USWDS tokens instead of CSS in JS.

clmedders commented:
Looks good overall... most of what i was going to point out james has already said.. i do have some design questions/thoughts 

for mobile, it might be good to have the menu take up the full screen and not allow the user to scroll as you can see when its open you can still scroll it and see some of the content 

<img width="523" alt="Screenshot 2024-04-24 at 10 32 00 AM" src="https://github.com/GSA/digitalgov.gov/assets/88721460/881f8ca7-2d89-45ff-8a37-27904d5e382e">

Also on mobile... the links are working however they are just taking the user to the top of the page.. i wonder if we should link to the start of the content like target the `id="content-start"` so that way the user it right at the section they have selected. 

On tablet its hard to know that there is more to the menu like no indication that if i scroll then i will have more options. Just something i noticed.. in the img below there is no way of knowing there is one more option in the bar 

<img width="669" alt="Screenshot 2024-04-24 at 10 34 36 AM" src="https://github.com/GSA/digitalgov.gov/assets/88721460/4b78abcb-7d9c-412d-8028-b310837e3e1e">

nick-mon1 commented on themes/digital.gov/layouts/partials/core/guides/guide-menu-links.html:
```diff
@@ -0,0 +1,36 @@
+        {{- end -}}
+      {{- end -}}
+    {{- end -}}
+    <a
+      class="{{ cond $current "dg-current" "" }}"
+      href="{{- $page.Permalink -}}#content-start"
+      title="{{- $item.title -}}"
+      >{{- $item.title -}}</a
```
Removed BEM class to allow easier overrides from the parent component.

nick-mon1 commented on themes/digital.gov/layouts/partials/core/guides/guide-menu-links.html:
```diff
@@ -0,0 +1,36 @@
+    {{- end -}}
+    <a
+      class="{{ cond $current "dg-current" "" }}"
+      href="{{- $page.Permalink -}}#content-start"
```
@clmedders Implemented your suggestion for navigating to top of content and not top of page.

nick-mon1 commented on themes/digital.gov/src/js/guide-mobile-menu.js:
Refactored to use classes instead of CSS-in-JS.

nick-mon1 commented on themes/digital.gov/src/scss/new/guides/_guide-menu-bar.scss:
```diff
@@ -54,6 +43,23 @@
     display: none;
   }
 
+  &-links a {
```
Override `<a>` tags for desktop or mobile styles.

mejiaj replied:
We should add a class to these links if possible.

I see the mobile nav is rendering empty class attribute:
![image](https://github.com/GSA/digitalgov.gov/assets/3385219/5210cb8b-9b9d-4267-8223-830b3dc8b955)

nick-mon1 commented:
New changes include:

- Used the hugo inline partial to be shared for both desktop and mobile menus
- Added documentation to updated guide partials
- Removed un-necessary javascript
- Flattened CSS code
- Switched the button to be on right side, will be easier for most users to press with thumb
- Tested `usa-accordion` against [USWDS accordion guidelines](https://designsystem.digital.gov/components/accordion/accessibility-tests/)

bonnieAcameron reviewed (changes requested):
I see the "glossary" icon twice--once in the banner near the top and again in the mobile menu. Can we remove the glossary icon from the mobile menu, which would also serve another purpose to reduce confusion about whether or not the book icon will open sections?*

*discussed with Nick in person, please disregard this suggestion. added my approval

mejiaj commented on themes/digital.gov/layouts/partials/core/guides/guide-menu-links.html:
```diff
@@ -0,0 +1,36 @@
+      {{- range $item.subnav -}}
+        {{- $subItem := . -}}
+        {{- $subPage := $pageScope.Site.GetPage $subItem.path -}}
+        {{- if eq $currentPage $subPage.RelPermalink -}}
+          {{- $current = true -}}
+          {{ printf "%#v" $current }}
+        {{- end -}}
```
Do we need this print statement? Not seeing this statement render anything.

nick-mon1 replied:
Removed, leftover from testing.

mejiaj commented on themes/digital.gov/src/js/guide-mobile-menu.js:
```diff
@@ -0,0 +1,43 @@
+const dgAccordionContent = document.querySelector(
+  ".guide-mobile-menu .usa-accordion__content"
+);
+const desktopBreakpoint = window.matchMedia("(min-width: 480px)");
```
The name `desktopBreakpoint` and its min-width being `480px`, does this mean that desktop starts at 480+?

nick-mon1 replied:
Yes, should it read `mobileBreakpoint` or `tabletBreakpoint` instead?

mejiaj commented on themes/digital.gov/src/js/guide-mobile-menu.js:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Display horizontal desktop or vertical mobile menu when viewport size changes
+ *
```
You can delete this empty line 3.

mejiaj commented on themes/digital.gov/src/js/guide-mobile-menu.js:
```diff
@@ -0,0 +1,43 @@
+
+const dgGuideMenuBar = document.querySelector(".dg-guide__menu-bar");
+const dgAccordionMobileMenu = document.querySelector(".guide-mobile-menu");
+const dgAccordionContent = document.querySelector(
+  ".guide-mobile-menu .usa-accordion__content"
+);
```
You have the mobile menu, you could query that instead of the document. This is faster than querying the entire DOM again.


```suggestion
const dgAccordionContent = dgAccordionMobileMenu.querySelector(".usa-accordion__content");
```

nick-mon1 replied:
Good catch!

mejiaj commented on themes/digital.gov/src/js/guide-mobile-menu.js:
```diff
@@ -0,0 +1,43 @@
+  desktopBreakpoint.addEventListener("change", handleResize);
+}
+
+initGuideMobileMenu();
```
Because this is bundled it will init on all pages. Consider a check to see if `dgGuideMenuBar` exists before running any of the code.

Check console on homepage:
![image](https://github.com/GSA/digitalgov.gov/assets/3385219/b6c840c1-39bf-4598-96f2-df60c03b34c0)

nick-mon1 replied:
Great point, wrapped in a conditional

mejiaj reviewed (changes requested):
Looks good. Noticing a visual quirk on click/active/focus. Added suggestions for improving code.

![image](https://github.com/GSA/digitalgov.gov/assets/3385219/6a48644e-bb24-4d51-97f4-9a23750740b6)

nick-mon1 commented:
@mejiaj 

Resolved comments above and addressed:

- added a class for the menu bar link: `<a class="dg-guide__menu-bar-link {{ cond $current "dg-current" "" }}">`
- wrapped the `guide-mobile-menu.js` in an IIFE and added conditional for checking if on a guide page
- removed the `:hover` styles for adding a `border-bottom` and also centered the text within the outline box on click.
<img width="430" alt="Screenshot 2024-05-16 at 2 42 31 PM" src="https://github.com/GSA/digitalgov.gov/assets/104778659/c6b10be6-dee0-4814-a0ca-ea87826d0edb">

- Fixed an icon flex layout bug when not on a guide page with a glossary

<img width="528" alt="Screenshot 2024-05-16 at 3 31 25 PM" src="https://github.com/GSA/digitalgov.gov/assets/104778659/cddbaf6c-692a-48c3-bdf4-2a1776016e38">

mejiaj commented on themes/digital.gov/src/scss/new/guides/_guide-menu-bar.scss:
```diff
@@ -12,6 +13,10 @@
   top: -1px;
   white-space: nowrap;
 
+  &--hide {
+    display: none;
+  }
```
Alternatively, you can use the `hidden` attribute.

nick-mon1 replied:
Set this to `visibility: hidden`

mejiaj commented on themes/digital.gov/src/scss/new/guides/_guide-menu-bar.scss:
```diff
@@ -54,6 +43,23 @@
   }
 
+  .dg-guide__menu-bar-link {
+    color: color("black");
```
Alternatively, ink is the default text color for USWDS.

```suggestion
    color: color("ink");
```

mejiaj reviewed (commented):
@nick-mon1 menu looks and works well. 

Added optional suggestions, feel free to resolve if they're beyond scope.

mejiaj reviewed (approved):
LGTM. Tested in Chromium and Firefox.
